storage/sqlite/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+use std::vec::IntoIter;
+
+use crate::error::process_sqlite_error;
+use storage_core::{
+    backend::{self, TransactionalRo, TransactionalRw},
+    info::DbDesc,
+    Data, DbIndex,
+};
+use utils::shallow_clone::ShallowClone;
+use utils::sync::Arc;
+
+/// Return the table name that corresponds to a specific index
+/// TODO this is unoptimised and the returned values should be cached
+fn db_table_name(idx: usize) -> String {
+    format!("db_{}", idx)
+}
+
+/// Identifiers of the list of databases (key-value maps)
+#[derive(Eq, PartialEq, Debug, Clone)]
+struct DbList(Vec<()>);
+
+impl std::ops::Index<DbIndex> for DbList {
+    type Output = ();
+
+    fn index(&self, index: DbIndex) -> &Self::Output {
+        &self.0[index.get()]
+    }
+}
+
+/// Sqlite iterator over entries with given key prefix
+pub struct PrefixIter {
+    /// Underlying iterator
+    iter: IntoIter<(Vec<u8>, Vec<u8>)>,
+
+    /// Prefix to iterate over
+    prefix: Data,
+}
+
+impl PrefixIter {
+    fn new(iter: IntoIter<(Vec<u8>, Vec<u8>)>, prefix: Data) -> Self {
+        PrefixIter { iter, prefix }
+    }
+}
+
+impl Iterator for PrefixIter {
+    type Item = (Data, Data);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let kv = self.iter.next()?;
+        utils::ensure!(kv.0.starts_with(&self.prefix));
+        Some(kv)
+    }
+}
+
+pub struct DbTx<'m> {
+    connection: MutexGuard<'m, Connection>,
+}
+
+impl<'m> DbTx<'m> {
+    pub fn start_transaction(connection: MutexGuard<'m, Connection>) -> storage_core::Result<Self> {
+        let tx = DbTx { connection };
+        tx.connection.execute("BEGIN TRANSACTION", ()).map_err(process_sqlite_error)?;
+        Ok(tx)
+    }
+
+    pub fn commit_transaction(&self) -> storage_core::Result<()> {
+        let _res = self
+            .connection
+            .execute("COMMIT TRANSACTION", ())
+            .map_err(process_sqlite_error)?;
+        Ok(())
+    }
+}
+
+impl Drop for DbTx<'_> {
+    fn drop(&mut self) {
+        if self.connection.is_autocommit() {
+            return;
+        }
+
+        let res = self.connection.execute("ROLLBACK TRANSACTION", ());
+        if let Err(err) = res {
+            println!("Error: transaction rollback failed: {}", err);
+        }
+    }
+}
+
+impl<'s, 'i> backend::PrefixIter<'i> for DbTx<'s> {
+    type Iterator = PrefixIter;
+
+    fn prefix_iter<'t: 'i>(
+        &'t self,
+        idx: DbIndex,
+        prefix: Data,
+    ) -> storage_core::Result<Self::Iterator> {
+        // TODO check if prefix.is_empty()
+        // TODO Perform the filtering in the SQL query itself
+        let mut stmt = self
+            .connection
+            .prepare_cached(
+                format!(
+                    "SELECT key, value FROM {} ORDER BY key",
+                    db_table_name(idx.get())
+                )
+                .as_str(),
+            )
+            .map_err(process_sqlite_error)?;
+
+        let mut rows = stmt.query(()).map_err(process_sqlite_error)?;
```

Comment:
It may be more convenient to use [`Statement::query_map`](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.query_map) here, it supposedly gives a standard Rust iterator. Have you explored that option?

Reply:
PrefixIter needs to be optimized so this will change